### PR TITLE
Support running tests in parallel

### PR DESCRIPTION
**Pull request by AdamGleave**
This PR reduces the time taken to run the test suite from ~45 minutes to ~8 minutes (when running 8 workers in parallel; further speed-up possible on sufficiently powerful machines). This required fixing some issues with non-deterministic ordering of parametrized tests (by fixing the Python hash ordering), and avoiding race conditions (writing to a fixed file name).

This does not speed up Travis CI builds, which are limited to only 2 cores. We might be able to do this by using a Travis test matrix, but I'm fine with the speed of CI builds, provided I can make sure things pass first in Docker on my local machine :)

Some of the bugs I found were non-deterministic so it's quite possible more gremlins will pop up; let me know if anything else that could cause race conditions comes to mind. However, the worst-case of a false positive in the test suite seems OK, as we can just fix it when it comes up.

### CODE REVIEW

**Comment by araffin** [comment]
Hey,
Sounds like a good feature to have ;)
Btw, how much does it take ? (That what i m mostly afraid of, also for travis)

Note: i need to update the docker image, so travis will use pytest-dist (the image is not built each time)

**Comment by AdamGleave** [comment]
You have to tell pytest to use `pytest-xdist` with `pytest -n NUM_JOBS`. Travis machines only have 2 cores so I'm not sure how much speed-up we'll see there, although we can try `pytest -n 2`.

File: tests/test_save.py [hunk 1]
```diff
@@ -127,5 +129,5 @@ def test_model_manipulation(model_class, storage_method):
         del model, env
 
     finally:
-        if os.path.exists("./test_model"):
-            os.remove("./test_model")
+        if os.path.exists(model_fname):
+            os.remove(model_fname)
```
Comment (araffin) [inline tests/test_save.py]: Shoud be `'{}.pkl'.format(model_fname)` no ?

**Comment by araffin** [comment]
>Travis machines only have 2 cores so I'm not sure how much speed-up 

I am more afraid of the memory issue... (I think we have some memory leaks that make travis fail sometimes, so with 2 processes, it would be even more likely :/ )

File: tests/test_save.py (hunk 1 above)
Comment (AdamGleave) [reply tests/test_save.py]: Good catch (I think this was a problem before this PR too). I've now changed `model_fname` everywhere to end in `.pkl` (`model.save` will not add a `.pkl` if this already exists, so the filename can be consistent throughout.)

**Comment by AdamGleave** [comment]
Ah I see. I think if we want to speed up Travis we should use build matrices with different subsets of the test-suite, not `pytest-xdist`.

**Comment by AdamGleave** [comment]
Bump @araffin. Are there any changes you want to see to this PR?

**Comment by araffin** [comment]
I think I just wanted to know @hill-a thoughts on that (in case I missed something), otherwise LGTM ;)

**Comment by araffin** [comment]
@AdamGleave I think we just need to check if there will be no problems with the new tests added to master, otherwise, we can merge ;)

**Comment by AdamGleave** [comment]
Passing on Travis (without xdist) and on my local machine in Docker (with xdist) so merging.

**Pull request merged**